snuba/subscriptions/executor_consumer.py

```diff
@@ -90,3 +102,127 @@ def create(
         self, commit: Callable[[Mapping[Partition, Position]], None]
     ) -> ProcessingStrategy[KafkaPayload]:
         return Noop(commit)
+
+
+class SubscriptionResultFuture(NamedTuple):
+    message: Message[KafkaPayload]
+    future: Future[Tuple[Request, Result]]
+
+
+class ExecuteQuery(ProcessingStrategy[KafkaPayload]):
+    """
+    Decodes a scheduled subscription task from the Kafka payload, builds
+    the request and executes the ClickHouse query.
+    """
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        max_concurrent_queries: int,
+        metrics: MetricsBackend,
+        next_step: ProcessingStrategy[KafkaPayload],
+    ) -> None:
+        self.__dataset = dataset
+        self.__max_concurrent_queries = max_concurrent_queries
+        self.__metrics = metrics
+        self.__next_step = next_step
+
+        self.__encoder = SubscriptionScheduledTaskEncoder()
+
+        self.__executor = ThreadPoolExecutor()
+        self.__queue: Deque[SubscriptionResultFuture] = deque()
+
+        self.__closed = False
+
+        self.__concurrent_gauge: Gauge = ThreadSafeGauge(
+            self.__metrics, "executor.concurrent"
+        )
+
+        self.__concurrent_clickhouse_gauge: Gauge = ThreadSafeGauge(
+            self.__metrics, "executor.concurrent.clickhouse"
+        )
+
+    def __execute_query(
+        self, task: ScheduledSubscriptionTask, tick: Tick
+    ) -> Tuple[Request, Result]:
+        timer = Timer("query")
+
+        with self.__concurrent_gauge:
+            request = task.task.subscription.data.build_request(
+                self.__dataset,
+                task.timestamp,
+                tick.offsets.upper,
+                timer,
+                self.__metrics,
+            )
+
+            result = parse_and_run_query(
+                self.__dataset,
+                request,
+                timer,
+                robust=True,
+                concurrent_queries_gauge=self.__concurrent_clickhouse_gauge,
+            ).result
+
+            return (request, result)
+
+    def poll(self) -> None:
+        while self.__queue:
+            if not self.__queue[0].future.done():
+                break
+
+            result_future = self.__queue.popleft()
+
+            message = result_future.message
+
+            self.__next_step.submit(message)
+
+        self.__next_step.poll()
+
+    def submit(self, message: Message[KafkaPayload]) -> None:
+        assert not self.__closed
+
+        # Tell the consumer to pause until we have removed some futures from
+        # the queue
+        if len(self.__queue) >= self.__max_concurrent_queries:
```

Comment:
I don't think you need to implement this check by yourself. 
Just create an executor pool with a number of thread with the maximum amount of concurrent threads you want. 
That enforces the maximum concurrently. 
Then you may also add a higher max queue size to apply back pressure when needed.

Reply:
Alright